uportal-war/src/main/java/org/jasig/portal/security/provider/AnyUnblockedGrantPermissionPolicy.java

```diff
@@ -65,161 +72,211 @@
+            final IPermissionOwner allPermissionsOwner = permissionOwnerDao.getPermissionOwner(IPermission.PORTAL_SYSTEM);
+            final IPermissionTarget allPermissionsTarget = targetProviderRegistry.getTargetProvider(allPermissionsActivity.getTargetProviderKey()).getTarget(IPermission.ALL_TARGET);
+            if (doesPrincipalHavePermission(service, principal, allPermissionsOwner, allPermissionsActivity, allPermissionsTarget)) {
+                // Stop checking;  just return true
+                return true;
             }
-            return true;
         }
 
-        // if the target is formatted as a channel, check if the user has
-        // the ALL_CHANNELS permission
-        if (target.startsWith(IPermission.PORTLET_PREFIX)
-                    && doesPrincipalHavePermission(service, principal, owner,
-                            activity, IPermission.ALL_PORTLETS_TARGET)) {
-            return true;
+        /*
+         * uPortal uses a few "special" targets that signal permission to
+         * perform the specified activity over an entire class of targets;
+         * see if one of those applies in this case.
+         */
+        IPermissionTarget collectiveTarget = null;  // The "collective noun" representing a class of thing
+        switch (target.getTargetType()) {
+            case PORTLET:
+                collectiveTarget = targetProviderRegistry.getTargetProvider(activity.getTargetProviderKey()).getTarget(IPermission.ALL_PORTLETS_TARGET);
+                break;
+            case CATEGORY:
+                collectiveTarget = targetProviderRegistry.getTargetProvider(activity.getTargetProviderKey()).getTarget(IPermission.ALL_CATEGORIES_TARGET);
+                break;
+            case GROUP:
+                collectiveTarget = targetProviderRegistry.getTargetProvider(activity.getTargetProviderKey()).getTarget(IPermission.ALL_GROUPS_TARGET);
+                break;
+            default:
+            // This sort of handling does not apply;  just pass through
         }
-
-        // if this target corresponds to a group or category, check if the user
-        // has the ALL_CATEGORIES or ALL_GROUPS permissions
-        IEntityGroup targetGroup = GroupService.findGroup(target);
-        if (targetGroup != null) {
-            if ((targetGroup.getEntityType().equals(IPortletDefinition.class)
-                    && doesPrincipalHavePermission(service, principal, owner,
-                            activity, IPermission.ALL_CATEGORIES_TARGET) || doesPrincipalHavePermission(
-                    service, principal, owner, activity,
-                    IPermission.ALL_GROUPS_TARGET))) {
+        /*
+         * NOTE:  Cannot generalize to a collective target if we are already on
+         * the collective target, else StackOverflowError.
+         */
+        if (collectiveTarget != null && !collectiveTarget.equals(target)) {
```

Comment:
Missed the above comment first glance at this line. Comment helps clarify the second part of the 'if' test -- thanks!